addon_service/oauth1/utils.py

```diff
@@ -0,0 +1,148 @@
+
+
+async def get_request_token(
+    request_token_url: str,
+    oauth_consumer_key: str,
+    oauth_consumer_secret: str,
+) -> tuple[OAuth1TokenCredentials, dict]:
+    """
+    Obtaining unauthorised request token needed to construct Authorization url
+    https://oauth.net/core/1.0a/#auth_step1
+    """
+    signed_headers = construct_signed_headers(
+        request_token_url, oauth_consumer_key, oauth_consumer_secret
+    )
+    return await _get_token(request_token_url, signed_headers)
+
+
+async def get_access_token(
+    access_token_url: str,
+    oauth_consumer_key: str,
+    oauth_consumer_secret: str,
+    oauth_token: str,
+    oauth_token_secret: str,
+    oauth_verifier: str,
+) -> tuple[OAuth1TokenCredentials, dict]:
+    """
+    Getting final access token needed to access protected resources from Service provider
+
+    """
+    signed_headers = construct_signed_headers(
+        access_token_url,
+        oauth_consumer_key,
+        oauth_consumer_secret,
+        oauth_token=oauth_token,
+        oauth_verifier=oauth_verifier,
+        oauth_token_secret=oauth_token_secret,
+    )
+    return await _get_token(access_token_url, signed_headers)
+
+
+async def _get_token(
+    request_token_url, signed_headers
+) -> tuple[OAuth1TokenCredentials, dict]:
+    _client = await get_singleton_client_session()
+    async with _client.post(
+        request_token_url, headers=signed_headers
+    ) as _token_response:
+        if not HTTPStatus(_token_response.status).is_success:
+            raise RuntimeError(await _token_response.text())
+        try:
+            return OAuth1TokenCredentials.from_dict(await _token_response.json())
+        except ContentTypeError:
+            raw_result = parse_qs(await _token_response.text())
+            result_dict = {key: value[0] for key, value in raw_result.items() if value}
+            return OAuth1TokenCredentials.from_dict(result_dict)
+
+
+def _construct_params(params_to_encode: dict) -> str:
+    return ",".join(
+        [f'{key}="{value}"' for key, value in sorted(params_to_encode.items())]
+    )
+
+
+def construct_signed_headers(
+    url: str,
+    oauth_consumer_key: str,
+    oauth_consumer_secret: str,
+    http_method: HTTPMethod = HTTPMethod.POST,
+    oauth_token: str | None = None,
+    oauth_token_secret: str | None = None,
+    oauth_verifier: str | None = None,
+):
+    oauth_params = construct_headers(oauth_consumer_key, oauth_token, oauth_verifier)
+    signature = generate_signature(
+        http_method, url, oauth_params, oauth_consumer_secret, oauth_token_secret
+    )
+    oauth_params |= {"oauth_signature": signature}
+    return {"Authorization": f"OAuth {_construct_params(oauth_params)}"}
+
+
+def generate_signature(
+    http_method: HTTPMethod,
+    url: str,
+    headers: dict,
+    oauth_consumer_secret: str,
+    oauth_token_secret: str | None = None,
+) -> str:
+    params_str = "&".join([f"{key}={headers[key]}" for key in sorted([*headers])])
```

Comment:
```suggestion
    params_str = "&".join(f"{key}={value}" for key, value in sorted(headers.items()))
```
nits: can skip the intermediate list and simplify as in `_construct_params` above